### PR TITLE
`NonNegativeParameterization` layer

### LINKED ISSUE #86: `NonNegativeParameterization` layer
Prerequisites

- [x] #44

### PR DESCRIPTION
**Pull request by 0x00b1**
closes #86

Non-negative parameterization as required by generalized divisive normalization (GDN) activations. The parameter is subjected to an invertible transformation that slows down the learning rate for small values.

A brief usage example:

```Python
import torch
from torch.nn import Module, Parameter
import torch.nn.functional

from torch import Tensor

from ._non_negative_parameterization import NonNegativeParameterization


class GeneralizedDivisiveNormalization(Module):
    def __init__(
        self,
        in_channels: int,
        inverse: bool = False,
        beta_min: float = 1e-6,
        gamma_init: float = 0.1,
    ):
        super(GeneralizedDivisiveNormalization, self).__init__()

        self._inverse = inverse

        self._reparameterized_beta = NonNegativeParameterization(
            torch.ones(in_channels),
            minimum=beta_min,
        )

        self._beta = Parameter(
            self._reparameterized_beta.initialized,
        )

        self._reparameterized_gamma = NonNegativeParameterization(
            gamma_init * torch.eye(in_channels),
        )

        self._gamma = Parameter(
            self._reparameterized_gamma.initialized,
        )

    def forward(self, x: Tensor) -> Tensor:
        _, channels, _, _ = x.size()

        y = torch.nn.functional.conv2d(
            x ** 2,
            torch.reshape(
                self._reparameterized_gamma(self._gamma),
                (channels, channels, 1, 1)
            ),
            self._reparameterized_beta(self._beta),
        )

        if self._inverse:
            return x * torch.sqrt(y)

        return x * torch.rsqrt(y)
```

```Python
import torch
import torch.testing

from neuralcompression.layers import GeneralizedDivisiveNormalization


class TestGeneralizedDivisiveNormalization:
    def test_backward(self):
        x = torch.rand((1, 32, 16, 16), requires_grad=True)

        generalized_divisive_normalization = GeneralizedDivisiveNormalization(32)

        y = generalized_divisive_normalization(x)

        y.backward(x)

        assert y.shape == x.shape

        assert x.grad is not None

        assert x.grad.shape == x.shape

        torch.testing.assert_allclose(
            x / torch.sqrt(1 + 0.1 * (x ** 2)),
            y,
        )

        generalized_divisive_normalization = GeneralizedDivisiveNormalization(
            32,
            inverse=True,
        )

        y = generalized_divisive_normalization(x)

        y.backward(x)

        assert y.shape == x.shape

        assert x.grad is not None

        assert x.grad.shape == x.shape

        torch.testing.assert_allclose(
            x * torch.sqrt(1 + 0.1 * (x ** 2)),
            y,
        )
```

### CODE REVIEW

**Review by mmuckley — APPROVED** [review]
All my comments addressed.
